contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,377 @@
+}
+
+// VolumeSnapshotContentList is a list of VolumeSnapshotContent objects
+type VolumeSnapshotContentList struct {
+        metav1.TypeMeta `json:",inline"`
+        Metadata        metav1.ListMeta      `json:"metadata"`
+        Items           []VolumeSnapshotContent `json:"items"`
+}
+
+// The desired state of the volume snapshot data
+type VolumeSnapshotContentSpec struct {
+	// Source represents the location and type of the volume snapshot
+	VolumeSnapshotSource `json:",inline" protobuf:"bytes,1,opt,name=volumeSnapshotSource"`
+
+	// VolumeSnapshotRef is part of bi-directional binding between VolumeSnapshot
+	// and VolumeSnapshotContent
+	// +optional
+	VolumeSnapshotRef *core_v1.ObjectReference `json:"volumeSnapshotRef" protobuf:"bytes,2,opt,name=volumeSnapshotRef"`
+
+	// PersistentVolumeRef represents the PersistentVolume that the snapshot has been
+	// taken from
+	// +optional
+	PersistentVolumeRef *core_v1.ObjectReference `json:"persistentVolumeRef" protobuf:"bytes,3,opt,name=persistentVolumeRef"`
+}
+
+// Represents the actual location and type of the snapshot.
+// Only CSI volume snapshot source is supported now.
+type VolumeSnapshotSource struct {
+	// CSI (Container Storage Interface) represents storage that handled by an external CSI Volume Driver (Alpha feature).
+	// +optional
+	CSI *CSIVolumeSnapshotSource `json:"csiVolumeSnapshotSource,omitempty"`
+}
+
+// Represents the source from CSI volume snapshot
+type CSIVolumeSnapshotSource struct {
+        // Driver is the name of the driver to use for this snapshot.
+        // Required.
+        Driver string `json:"driver"`
+
+        // SnapshotHandle is the unique snapshot id returned by the CSI volume
+        // plugin’s CreateSnapshot to refer to the snapshot on all subsequent calls.
+        // Required.
+        SnapshotHandle string `json:"snapshotHandle"`
+
+        // Timestamp when the point-in-time snapshot is taken on the storage
+        // system. The format of this field should be a Unix nanoseconds time
+        // encoded as an int64. On Unix, the command `date +%s%N` returns
+        // the  current time in nanoseconds since 1970-01-01 00:00:00 UTC.
+        // This field is REQUIRED.
+	CreatedAt int64 `json:"createdAt,omitempty" protobuf:"varint,3,opt,name=createdAt"`
+}
+
+```
+
+#### The `DataSource` Object in PVC
+
+Add a new field into PVC to represent the source of the data which is prepopulated to the provisioned volume. Possible source types may include the following:
+
+    * VolumeSnapshot: restore snapshot to a new volume
+    * PersistentVolumeClaim: clone volume which is represented by PVC
+
+```
+
+type PersistentVolumeClaimSpec struct {
+        // If specified, volume will be prepopulated with data from the PVCDataSourceRef.
+        // +optional
+        PVCDataSourceRef *core_v1.LocalObjectReference `json:"dataSourceRef" protobuf:"bytes,2,opt,name=dataSourceRef"`
+
+}
```

Comment:
Would a v1.ObjectReference be better? This also allows sources to be in other namespaces.

Reply:
The intention is to only allow sources in the same namespace to be used.

Reply:
Why is this restriction in place? Is it a technical reason or something else?

Reply:
It is for technical reason.  VolumeSnapshot itself is name-spaced.